ckanext/cloudstorage/logic/action/multipart.py

```diff
@@ -0,0 +1,221 @@
+        if ~_rindex:
+            try:
+                name_prefix = res_name[:_rindex]
+                for cloud_object in uploader.container.iterate_objects():
+                    if cloud_object.name.startswith(name_prefix):
+                        log.info('Removing cloud object: %s' % cloud_object)
+                        cloud_object.delete()
+            except Exception as e:
+                log.error('[delete from cloud] %s' % e)
+
+        resp = uploader.driver.connection.request(
+            _get_object_url(uploader, res_name) + '?uploads',
+            method='POST'
+        )
+        if not resp.success():
+            raise toolkit.ValidationError(resp.error)
+        try:
+            upload_id = resp.object.find(
+                '{%s}UploadId' % resp.object.nsmap[None]).text
+        except AttributeError:
+            upload_id_list = filter(
+                lambda e: e.tag.endswith('UploadId'),
+                resp.object.getchildren()
+            )
+            upload_id = upload_id_list[0].text
+        upload_object = MultipartUpload(upload_id, id, res_name, size, name)
+
+        upload_object.save()
+    return upload_object.as_dict()
+
+
+def upload_multipart(context, data_dict):
+    h.check_access('cloudstorage_upload_multipart', data_dict)
+    upload_id, part_number, part_content = toolkit.get_or_bust(
+        data_dict, ['uploadId', 'partNumber', 'upload'])
+
+    uploader = ResourceCloudStorage({})
+    upload = model.Session.query(MultipartUpload).get(upload_id)
+
+    resp = uploader.driver.connection.request(
+        _get_object_url(
+            uploader, upload.name) + '?partNumber={0}&uploadId={1}'.format(
+                part_number, upload_id),
+        method='PUT',
+        data=bytearray(part_content.file.read())
+    )
+    if resp.status != 200:
+        raise toolkit.ValidationError('Upload failed: part %s' % part_number)
+
+    _save_part_info(part_number, resp.headers['etag'], upload)
+    return {
+        'partNumber': part_number,
+        'ETag': resp.headers['etag']
+    }
+
+
+def finish_multipart(context, data_dict):
+    """Called after all parts had been uploaded.
+
+    Triggers call to `_commit_multipart` which will convert separate uploaded parts into single file
+
+    :param context:
+    :param data_dict:
+        dict with required key `uploadId` - id of Multipart Upload that should be finished
+    :returns: None
+    :rtype: NoneType
+
+    """
+
+    h.check_access('cloudstorage_finish_multipart', data_dict)
+    upload_id = toolkit.get_or_bust(data_dict, 'uploadId')
+    upload = model.Session.query(MultipartUpload).get(upload_id)
+    chunks = [
+        (part.n, part.etag)
+        for part in model.Session.query(MultipartPart).filter_by(
+            upload_id=upload_id).order_by(MultipartPart.n)
+    ]
+    uploader = ResourceCloudStorage({})
+    try:
+        obj = uploader.container.get_object(upload.name)
+        obj.delete()
+    except:
```

Comment:
Never use a blind exception, because this catches and throws away things like `^C`. Use `except Exception`.